File: acwing/算法提高课/第一章动态规划/482.合唱队形/482.合唱队形.cpp
#include <iostream>
#include <cstring>
#include <algorithm>

using namespace std;

const int N = 110;

int f[N], g[N], a[N];

int main()
{
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++)
        cin >> a[i];

    for (int i = 1; i <= n; i++)
    {
        f[i] = 1;
        for (int j = 1; j < i; j++)
        {
            if (a[i] > a[j])
                f[i] = max(f[i], f[j] + 1);
        }
    }
    for (int i = n; i >= 1; i--)
    {
        g[i] = 1;
        for (int j = n; j > i; j--)
        {
            if (a[i] > a[j])
                g[i] = max(g[i], g[j] + 1);
        }
    }
    int res = 0;
    for (int k = 1; k <= n; k++)
        res = max(res, f[k] + g[k] - 1);

    printf("%d\n", n - res);
    return 0;
}

File: acwing/算法提高课/第一章动态规划/482.合唱队形/482.合唱队形.py
N = 110

f = [0] * N
g = [0] * N

a = []

n = int(input())

a = [0] + list(map(int, input().split()))

for i in range(1, n + 1):
    f[i] = 1
    for j in range(1, i):
        if a[i] > a[j]:
            f[i] = max(f[i], f[j] + 1)

for i in range(n, 0, - 1):
    g[i] = 1 
    for j in range(n, i, -1):
        if a[i] > a[j]:
            g[i] = max(g[i], g[j] + 1)
res = 0
for i in range(1, n + 1):
    res = max(res, f[i] + g[i] - 1)

print(n - res)
